cmd/zrepl/zrepl.c

```diff
@@ -64,122 +65,180 @@ uzfs_zvol_read_header(int fd, zvol_io_hdr_t *hdr)
+		return (-1);
+	}
+	if (hdr.opcode != ZVOL_OPCODE_OPEN) {
+		ZREPL_ERRLOG("Zvol must be opened first\n");
+		return (-1);
+	}
+	if (hdr.len != sizeof (open_data)) {
+		ZREPL_ERRLOG("Invalid payload length for open\n");
+		return (-1);
+	}
+	rc = uzfs_zvol_socket_read(fd, (char *)&open_data, sizeof (open_data));
+	if (rc != 0) {
+		ZREPL_ERRLOG("Payload read failed: %d\n", errno);
+		return (-1);
+	}
+
+	open_data.volname[MAX_NAME_LEN - 1] = '\0';
+	zinfo = uzfs_zinfo_lookup(open_data.volname);
+	if (zinfo == NULL) {
+		ZREPL_ERRLOG("zvol %s not found", open_data.volname);
+		hdr.status = ZVOL_OP_STATUS_FAILED;
+		goto open_reply;
+	}
+	zv = zinfo->zv;
+	ASSERT3P(zv, !=, NULL);
+	if (zv->zv_metavolblocksize != 0 &&
+	    zv->zv_metavolblocksize != open_data.tgt_block_size) {
+		ZREPL_ERRLOG("Conflicting block size");
+		hdr.status = ZVOL_OP_STATUS_FAILED;
+		goto open_reply;
+	}
+	// validate block size (only one bit is set in the number)
+	if (open_data.tgt_block_size == 0 ||
+	    (open_data.tgt_block_size & (open_data.tgt_block_size - 1)) != 0) {
+		ZREPL_ERRLOG("Invalid block size");
+		hdr.status = ZVOL_OP_STATUS_FAILED;
+		goto open_reply;
+	}
+
+	(void) pthread_mutex_lock(&zinfo->zinfo_mutex);
+	/*
+	 * Hold objset if this is the first query for the zvol. This can happen
+	 * in case that the target creates data connection directly without
+	 * getting the endpoint through mgmt connection first.
+	 */
+	if (zv->zv_objset == NULL && uzfs_hold_dataset(zv) != 0) {
+		(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
+		ZREPL_ERRLOG("Failed to hold zvol during open\n");
+		hdr.status = ZVOL_OP_STATUS_FAILED;
+		goto open_reply;
+	}
+	if (uzfs_update_metadata_granularity(zv,
+	    open_data.tgt_block_size) != 0) {
+		(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
+		ZREPL_ERRLOG("Failed to set granularity of metadata\n");
+		hdr.status = ZVOL_OP_STATUS_FAILED;
+		goto open_reply;
+	}
+	/*
+	 * TODO: Once we support multiple concurrent data connections for a
+	 * single zvol, we should probably check that the timeout is the same
+	 * for all data connections.
+	 */
+	uzfs_update_ionum_interval(zinfo, open_data.timeout);
+	zinfo->timeout = open_data.timeout;
+	*zinfopp = zinfo;
+
+	if (!zinfo->is_io_ack_sender_created) {
+		thrd_arg = kmem_alloc(sizeof (thread_args_t), KM_SLEEP);
+		thrd_arg->fd = fd;
+		thrd_arg->zinfo = zinfo;
+		uzfs_zinfo_take_refcnt(zinfo, B_FALSE);
+		zinfo->conn_closed = B_FALSE;
+		zinfo->is_io_ack_sender_created = B_TRUE;
+		thrd_info = zk_thread_create(NULL, 0,
+		    (thread_func_t)uzfs_zvol_io_ack_sender,
+		    (void *)thrd_arg, 0, NULL, TS_RUN, 0,
+		    PTHREAD_CREATE_DETACHED);
+		VERIFY3P(thrd_info, !=, NULL);
+	}
+	(void) pthread_mutex_unlock(&zinfo->zinfo_mutex);
+
+	hdr.status = ZVOL_OP_STATUS_OK;
+
+open_reply:
+	hdr.len = 0;
+	rc = uzfs_zvol_socket_write(fd, (char *)&hdr, sizeof (hdr));
+	if (rc == -1)
+		ZREPL_ERRLOG("failed to send reply for open request\n");
```

Comment:
this should be considered as failure and need to return -1 after cleanup like decrementing refcount, killing threads etc

Reply:
good catch. I will fix that.